### ISSUE TITLE
directive's link function called twice

### ISSUE THREAD

**Issue by pulkitnandan** [angular/angular.js#12780]
I have a very simple example of d3js and angularjs which is calling its render function twice while I want it to call it only once and I don't understand why its been executed twice.

Here goes the plunker http://plnkr.co/edit/YoWsbJeWJgONb8rcMdpA?p=preview

**Comment by gkalpak** [comment]
That sounds like a general support question and as such should be directed to the [appropriate channels][1].
GitHub issues are reserved for bug reports and feature requests.

--
BTW, :+1: for including a live plnkr demo, but keep in mind that keeping the reproduction as short as possible (leaving only the minimum stuff necessary to reproduce the problem), dramatically increases the chances of getting a quick response (or any response at all).

[1]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#question

**Comment by gkalpak** [comment]
@pulkitnandan, I just got curious:

The reason it is called twice, is because you `$watch` the window's `innerWidth` and call `render()` in the watchAction (listener).

Quoting from the docs on [`$watch()`][1]:

> After a watcher is registered with the scope, the listener fn is called asynchronously (via $evalAsync) to initialize the watcher. In rare cases, this is undesirable because the listener is called when the result of watchExpression didn't change. To detect this scenario within the listener fn, you can compare the newVal and oldVal. If these two values are identical (===) then the listener was called due to initialization.

--
"Normally", I would advice to change the listener fn like this:

```js
scope.$watch(..., function (newVal, oldVal) {
    if (newVal !== oldVal) scope.render(...);
});
```

**BUT**...

...`$watch`ing DOM values (such as `window.innerWith`) is a bad practice (performance-wise), mainly because the watchExpression will might be called multiple times (at least 2) during **ever** digest cycle. Things get even worse if your code leads layout-thrashing...
So, you might want to change your approach altogether...

[1]: https://docs.angularjs.org/api/ng/type/$rootScope.Scope#$watch

**Comment by pulkitnandan** [comment]
So to debug angularjs isn't this a prerequisite to rote learn it's api doc's till throat and then even think of developing it. As googling for its solution never to me to that particular line of code and hail how do people debug angularjs, I remember putting up a debugger; at a point where I thought it might be deviating from what is expected but it take me 15 minutes to come out of angular code, it was looping some function insanely and each time all interations completed it started the iteration again, I could never reach user code.

**Comment by gkalpak** [comment]
@pulkitnandan, this is obviously out-of-scope here, but stepping through every statement of the frameworks intrnals sure isn't the best way to debug - it's better to use break-points on the code that interests you (e.g. put a break-point in your `render()` function and examine the stack-trace whenever it's called to see where the extra call is coming from).

Additionally, the documentation of `$watch()` (quoted [above]) does explain quite clearly what happens when you register a watcher (i.e. that it is executed immediately and how to handle this in case it is undesirable).

I don't really see what could be improved for this particular usecase (in terms of debugging), but if you have suggestions, then by all means open a new feature request (or PR :smiley:).

[1]: https://github.com/angular/angular.js/issues/12780#issuecomment-138500055
